### PR TITLE
MSM: Bring in quotient polynomial and constraints 

### PR DESCRIPTION
**Pull request by volhovm**
Solves #1790: 
1. Moves constraint creating logic into MSM prover/verifier
2. Hopefully wiring it up with extra simple circuit to prove it works

### CODE REVIEW

File: msm/src/prover.rs [hunk 1]
```diff
@@ -54,45 +79,84 @@ where
...
+        Some(LookupProof {
+            m: lookup_env.lookup_counters_comm_d1.clone(),
+            h: lookup_env.lookup_terms_comms_d1.clone(),
+            sum: lookup_env.lookup_aggregation_comm_d1.clone(),
+        })
+    } else {
+        None
+    };
+
+    // TODO absorb mvlookup_comms
```
Comment (dannywillems) [inline msm/src/prover.rs:109]: Note: we do absorb already in the LookupEnv

File: msm/src/prover.rs [hunk 2]
```diff
@@ -25,27 +29,48 @@ pub fn prove<
...
 {
+    ////////////////////////////////////////////////////////////////////////////
+    // Setting up the protocol
+    ////////////////////////////////////////////////////////////////////////////
+
+    // TODO: rng should be passed as arg
+    let rng = &mut rand::rngs::OsRng;
+
+    let max_poly_size: u64 = unimplemented!();
+    let num_chunks: usize = if domain.d1.size < max_poly_size {
```
Comment (dannywillems) [inline msm/src/prover.rs:46]: Not useful here.

File: msm/src/prover.rs [hunk 3]
```diff
@@ -54,45 +79,84 @@ where
...
+        None
+    };
+
+    // TODO absorb mvlookup_comms
+
     // -- end computing the running sum in lookup_aggregation
     // -- End of MVLookup
 
-    // TODO: add quotient polynomial (based on constraints and expresion framework)
+    ////////////////////////////////////////////////////////////////////////////
```
Comment (dannywillems) [inline msm/src/prover.rs:114]: Nice comments. Let's make a small PR tomorrow morning just adding unimplemented code.

File: msm/src/prover.rs [hunk 4]
```diff
@@ -105,21 +169,100 @@ where
...
+            public_evals: None,
+            witness_evals,
+            mvlookup_evals,
+        }
+    };
 
-    let mut polynomials: Vec<_> = polynomials
-        .iter()
+    // Challenges and constants used for evaluating expressions
+    let expr_chals = Challenges {
```
Comment (dannywillems) [inline msm/src/prover.rs:218]: Let's make a small PR adding these, built on top of https://github.com/o1-labs/proof-systems/pull/1850

**Comment by dannywillems** [comment]
Closing. See serialization subdirectory

**Comment by dannywillems** [comment]
@volhovm Can you fix the conflicts please?

**Comment by volhovm** [comment]
@dannywillems ah, oh well. on it.

**Comment by dannywillems** [comment]
Can you squash: https://github.com/o1-labs/proof-systems/pull/1839/commits/be3d64e918183a68383709a5894c913dd62cca8f and https://github.com/o1-labs/proof-systems/pull/1839/commits/bcd09ef0e93406c977f19322f8c2c0ca0ce41301 please? There are unecessary commits.

File: msm/src/prover.rs [hunk 5]
```diff
@@ -66,45 +84,159 @@ where
...
+
     // -- end computing the running sum in lookup_aggregation
     // -- End of MVLookup
 
-    // TODO: add quotient polynomial (based on constraints and expresion framework)
+    // TODO rename this
+    // The evaluations should be at least the degree of our expressions. Higher?
+    // Maybe we can only use d4, we don't have degree-7 gates anyway
+    let mut witness_evals_env: Vec<Evaluations<G::ScalarField, R2D<G::ScalarField>>> = vec![];
+    for witness_poly in witness_polys.x.iter() {
```
Comment (dannywillems) [inline msm/src/prover.rs]: You should be able to simple use `iter`. We want to abstract the Witness structure.

**Comment by volhovm** [comment]
@dannywillems I know it's annoying but I can't really squash these without messing everything up sadly because I merged `master` into it, so now `git rebase -i` treats my commits and `master` commits equally. If you know a way to do it please tell me

**Comment by dannywillems** [comment]
https://github.com/o1-labs/proof-systems/pull/1839#issuecomment-1977236244

The only way I think is restarting the branch before the git merge, squash, redo git merge, git push force.

File: msm/src/prover.rs (hunk 5 above)
Comment (dannywillems) [reply msm/src/prover.rs]: See b8eb40e99c15572362213c3b3520034c0f62a54d.

File: msm/src/main.rs [hunk 6]
```diff
@@ -52,11 +56,14 @@ pub fn main() {
...
     println!("Verifying the proof");
     let verifies = verify::<_, OpeningProof, BaseSponge, ScalarSponge, MSM_FFADD_N_COLUMNS>(
-        domain, &srs, &proof,
+        domain,
+        &srs,
+        &constraint_exprs,
+        &proof,
     );
-    println!("Proof verification result: {verifies}")
+    println!("=================\n   Proof verification result: {verifies}")
```
Comment (dannywillems) [inline msm/src/main.rs:68]: Useful?

File: msm/src/prover.rs [hunk 7]
```diff
@@ -83,32 +103,163 @@ where
...
-        };
-        (evals(&zeta), evals(&zeta_omega))
+
+    // For initializitg arrays. Use MaybeUnitit instead:
+    // https://doc.rust-lang.org/core/mem/union.MaybeUninit.html#initializing-an-array-element-by-element
+    let stub_zero_eval: PointEvaluations<G::ScalarField> = PointEvaluations {
+        zeta: G::ScalarField::zero(),
+        zeta_omega: G::ScalarField::zero(),
+    };
+    let witness_evals = {
```
Comment (dannywillems) [inline msm/src/prover.rs:249]: You can use a map with into_iter

File: msm/src/constraint.rs [hunk 8]
```diff
@@ -1,3 +1,7 @@
+use ark_ff::Zero;
```
Comment (dannywillems) [inline msm/src/constraint.rs:1]: Let's move/clean this in the subdirectory in the next PR

File: msm/src/main.rs [hunk 9]
```diff
@@ -1,7 +1,10 @@
-use ark_ff::UniformRand;
```
Comment (dannywillems) [inline msm/src/main.rs:1]: As column.rs, let's move this in a subdirectory for ffadd/ffmul in the next PR. Would be nice to clean this

File: msm/src/prover.rs [hunk 10]
```diff
@@ -83,32 +103,163 @@ where
...
+        let Witness { cols } = &witness_polys;
+        let eval_foo = |point, poly: &DensePolynomial<G::ScalarField>| poly.evaluate(point);
```
Comment (dannywillems) [inline msm/src/prover.rs:251]: Not sure `eval_foo` is the best name :)

File: msm/src/constraint.rs (hunk 8 above)
Comment (volhovm) [reply msm/src/constraint.rs:1]: Moved in https://github.com/o1-labs/proof-systems/pull/1904

**Review by dannywillems — APPROVED** [review]
Let's just merge it. We'll fix later. We will iterate over it. I didn't test it, but we will after. It does create tons of conflicts in the next PRs.
Let's have very small PR later.

I would revert the changes in kimchi to decrease as much as possible the changes outside of the MSM code

**Pull request merged**